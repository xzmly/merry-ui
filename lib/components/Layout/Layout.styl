@import '../../helpers/stylus.helper.styl';

layout = class-prefix + "-layout"
header = class-prefix + "-layout-header"
content = class-prefix + "-layout-content"
footer = class-prefix + "-layout-footer"
aside = class-prefix + "-layout-aside"
has-aside = class-prefix + "-layout-has-aside"

header-height = 64px
footer-height = 64px
aside-width = 240px

.{layout},.{header},.{content},.{footer},.{aside}
  box-sizing border-box

.{layout}
  display flex
  flex-direction column
  height 100%
  &.{has-aside}
    flex-direction row
    > .{layout}
      flex 1
  > .{has-aside}
    flex 1
    flex-direction row

.{header}
  width 100%
  height header-height
  line-height: header-height
  box-shadow: 0 2px 8px #f0f1f2;

.{content}
  flex 1

.{footer}
  width 100%
  height footer-height

.{aside}
  width aside-width
