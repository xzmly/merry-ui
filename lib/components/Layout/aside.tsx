import React from "react";
import classes from "../../helpers/classes";

export interface AsideProps extends React.HTMLAttributes<HTMLDivElement>{
  className?: string
}

const Aside: React.FC<AsideProps> =
    props => {

      const {className,children,style} = props

      return (
          <aside style={style} className={classes(className, 'layout-aside')}>{children}</aside>
      )
    }

export default Aside
