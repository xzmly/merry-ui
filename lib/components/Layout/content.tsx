import React from "react";
import classes from "../../helpers/classes";

export interface ContentProps {
  className?: string
}

const Content: React.FC<ContentProps> =
    props => {

      const {className} = props

      return (
          <div className={classes(className, 'content')}>x</div>
      )
    }

export default Content
