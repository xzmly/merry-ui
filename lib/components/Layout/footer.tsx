import React from "react";
import classes from "../../helpers/classes";

export interface FooterProps extends React.HTMLAttributes<HTMLDivElement>{
  className?: string
}

const Footer: React.FC<FooterProps> =
    props => {

      const {className,children,style} = props

      return (
          <footer style={style} className={classes(className, 'layout-footer')}>
            {children}</footer>
      )
    };

export default Footer
