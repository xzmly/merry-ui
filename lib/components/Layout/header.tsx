import React from "react";
import classes from "../../helpers/classes";

export interface HeaderProps extends React.HTMLAttributes<HTMLDivElement>{
  className?: string
}

const Header: React.FC<HeaderProps> =
    props => {

      const {className,children,style} = props

      return (
          <header style={style} className={classes(className, 'layout-header')}>
            {children}</header>
      )
    };

export default Header
