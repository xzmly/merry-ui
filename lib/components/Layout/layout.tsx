import * as React from 'react';
import classes from '../../helpers/classes';
import Aside,{ AsideProps }from "./aside"
import './layout.styl';

interface LayoutProps extends React.HTMLAttributes<HTMLDivElement> {
  className?: string
}

interface ChildrenProps extends LayoutProps{
  typeName: 'header' | 'main' | 'footer'
  initClassName: string
}

export interface LayoutComponent<P> extends React.FC<P> {
  Header: React.ComponentClass<LayoutProps>
  Content: React.ComponentClass<LayoutProps>
  Footer: React.ComponentClass<LayoutProps>
  Aside: React.FC<AsideProps>
}


const Layout: LayoutComponent<LayoutProps> =
        props => {

  const { className,children,style } = props;

  const hasAside: string =
      children &&
      !(children instanceof Array) &&
      (children as any).type &&
      (children as any).type.name === 'Aside'
      ||
      children instanceof Array &&
      children.some((v:any) => v.type && v.type.name === 'Aside')
      ? "layout-has-aside" : '';


  return (
      <section style={style} className={classes(className,'layout',hasAside)}>
        {children}
      </section>
  )
};


const createComponent = (ExpectComponent: any,{typeName,initClassName}: ChildrenProps) =>
    class Component extends React.Component<ChildrenProps,any>{
      render(){
        const { className,children,...restProps } = this.props;
        return(
            <ExpectComponent typeName={typeName}
                             className={classes(className,`layout-${initClassName}`)}
                             children={children}
                             {...restProps}/>
        )
      }
    };

const createElement = (props: ChildrenProps) => {
  const { typeName,children,...restProps } = props;
  return React.createElement(typeName,{ ...restProps },children)
};

Layout.Header = createComponent(createElement,{
  typeName: 'header',
  initClassName: 'header'
});

Layout.Content = createComponent(createElement,{
  typeName: 'main',
  initClassName: 'content'
});

Layout.Footer = createComponent(createElement,{
  typeName: 'footer',
  initClassName: 'footer'
});

Layout.Aside = Aside;

export default Layout
