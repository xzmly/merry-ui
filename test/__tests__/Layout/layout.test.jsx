import renderer from 'react-test-renderer'
import Layout from '../../../lib/components/Layout/layout'
import React from 'react'
import {mount} from "enzyme/build/index";

const { Aside,Content, Header, Footer} = Layout;

describe('layout',()=> {
  it('layout 里如果有 aside 就会增加一个has-aside className',()=>{
    const c = mount(<Layout>
      <div>123</div>
      <Aside/>
    </Layout>);
    expect(c.find('section').hasClass('merry-layout-has-aside')).toBe(true)
  })
  it('是一个section', () => {
    const json = renderer.create(<Layout/>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('是一个header',()=>{
    const json = renderer.create(<Header/>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('是一个footer',()=>{
    const json = renderer.create(<Footer/>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('是一个main',()=>{
    const json = renderer.create(<Content/>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('是一个aside',()=>{
    const json = renderer.create(<Aside/>).toJSON()
    expect(json).toMatchSnapshot()
  })
});
