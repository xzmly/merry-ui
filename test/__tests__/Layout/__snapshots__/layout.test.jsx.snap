// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout 是一个aside 1`] = `
<aside
  className="merry-layout-aside"
/>
`;

exports[`layout 是一个footer 1`] = `
<footer
  className="merry-layout-footer"
/>
`;

exports[`layout 是一个header 1`] = `
<header
  className="merry-layout-header"
/>
`;

exports[`layout 是一个main 1`] = `
<main
  className="merry-layout-content"
/>
`;

exports[`layout 是一个section 1`] = `
<section
  className="merry-layout"
/>
`;
